### ISSUE TITLE
Add a "patch" source type

### ISSUE THREAD

**Issue by cs-shadow** [cs-shadow/buildstream#67]
In GitLab by @bochecha on Aug 7, 2017, 14:25

It's currently possible to add patches as a `local` source: https://gitlab.com/BuildStream/gnome-test-data/blob/48e30c94a85477566c5c77d2790545955d12c6c2/elements/core-deps/mozjs38.bst

It would be much nicer to be able to add `patch` sources, for example something like this:

```yaml
sources:
- kind: tar
  url: ...
  ref: ...
- kind: patch
  path: ...
  strip-components: 0
```

The `strip-components` would default to `1` (like in `flatpak-builder` manifests) as that's probably the most common these days given how everybody uses Git.

BuildStream would then automatically patch the sources with the `patch` sources, in the order they are declared.

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Aug 21, 2017, 19:17

This sounds like a sensible convenience; it could be a `Source` plugin which applies the patch at `Source.stage()` time and raises an error in case of staging into an empty directory.

Would have to raise an appropriate error if the user tries to open a workspace on a `patch` source. Handling patch workspaces in a custom way is probably not worthwhile; especially since this one would probably only support local patches (not remote URIs).

**Comment by cs-shadow** [comment]
In GitLab by @cs-shadow on Sep 7, 2017, 12:40

mentioned in merge request !86

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Sep 9, 2017, 22:11

closed via merge request !86

**Comment by cs-shadow** [comment]
In GitLab by @cs-shadow on Sep 9, 2017, 22:11

closed via commit b1b5cd56386272b04487eb7563198215ce9ae9a1
